internal/watcher/watcher_plugin.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+package watcher
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/nginx/agent/v3/internal/bus"
+	"github.com/nginx/agent/v3/internal/config"
+)
+
+type (
+	Watcher struct {
+		messagePipe            bus.MessagePipeInterface
+		agentConfig            *config.Config
+		instanceWatcherService *InstanceWatcherService
+		instanceUpdatesChannel chan InstanceUpdates
+		cancel                 context.CancelFunc
+	}
+)
+
+func NewWatcher(agentConfig *config.Config) *Watcher {
+	return &Watcher{
+		agentConfig:            agentConfig,
+		instanceWatcherService: NewInstanceWatcherService(agentConfig),
+		instanceUpdatesChannel: make(chan InstanceUpdates),
+	}
+}
+
+// nolint: unparam
+// error is always nil
+func (w *Watcher) Init(ctx context.Context, messagePipe bus.MessagePipeInterface) error {
+	slog.DebugContext(ctx, "Starting watcher plugin")
+	w.messagePipe = messagePipe
+
+	watcherContext, cancel := context.WithCancel(ctx)
+	w.cancel = cancel
+
+	go w.instanceWatcherService.Watch(watcherContext, w.instanceUpdatesChannel)
+	go w.monitorInstanceUpdates(watcherContext)
+
+	return nil
+}
+
+// nolint: unparam
+// error is always nil
+func (w *Watcher) Close(ctx context.Context) error {
+	slog.DebugContext(ctx, "Closing watcher plugin")
+
+	w.cancel()
+
+	return nil
+}
+
+func (*Watcher) Info() *bus.Info {
+	return &bus.Info{
+		Name: "watcher",
+	}
+}
+
+func (*Watcher) Process(_ context.Context, _ *bus.Message) {}
+
+func (*Watcher) Subscriptions() []string {
+	return []string{}
+}
+
+func (w *Watcher) monitorInstanceUpdates(ctx context.Context) {
```

Comment:
We may want this to be more generic so that we listen to updates from all watchers? Or do we want a unique consumer goroutine for each watcher?

Reply:
The channels will be of different types depending on the watcher. The instance watcher will return instance updates, file watcher will return file updates, etc. So I think its better to have a separate channel for each watcher unless we want to have a generic object for storing updates from watchers.

Reply:
But could we not listen to events from multiple different channels in the `select`? Regardless, maybe this can be added in later if we really want to.

Reply:
Oh yeah that makes sense. I think thats a good idea actually let me update that.